Assignment 2

dbhattacharyya-codenation opened the pull request:
Backup branch for assignment2 rebased off `master`

coderick14 committed:
Add code for assignment 2

- Isuue type 1 works completely
- Dockerized application

coderick14 committed:
Add README

coderick14 committed:
Populate initial data in MySQL on server startup

coderick14 committed:
Handle issueType 3

- Synchronize lazy initialization of static fields
- Add more modifiers for issueType 1

coderick14 committed:
Improve build instructions in README

coderick14 committed:
Fix logical bug in MinifixController

- Run update queries on Codegraph before hitting codegen

pranavram-cn commented on gradlew:
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env sh
```
Do not check in binaries and generated files into a VCS. USe .gitignore to make sure these aren't tracked.

pranavram-cn commented on src/main/java/simplefixer/HelloController.java:
```diff
@@ -0,0 +1,13 @@
+
+    @RequestMapping("/chal")
+    public String dumbMethod() {
+        return "Chal gaya bhai!!";
```
Congratulations!!

pranavram-cn commented on src/main/java/simplefixer/model/Minifix.java:
```diff
@@ -0,0 +1,116 @@
+
+    public Minifix() {}
+
+    public Integer getId() {
```
Getters and Setters, while necessary, add way too much noise to the codebase. It's recommended to use Lombok and its @Getter and @Setter annotations. Look into it.

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,403 @@
+import java.util.*;
+
+// Serializer class for post response
+class PostResponse {
```
Do not define classes within other files unless absolutely necessary. I recommend defining all these helper classes in a separate package and importing them here.

dbhattacharyya-codenation replied:
Fixed by [#0e87a7](https://github.com/dbhattacharyya-codenation/jcnu18/pull/1/commits/0e87a78060f441f0835dfe1be99b4dc2436591e5)

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,403 @@
+
+    // helper method to create neo4j database driver
+    private void createNeo4jDriver(String boltURL) {
+        String username = "neo4j";
```
Not recommended to use constants directly in code. Define a utility class called Constants or something like that and define all your constants there. Usernames, queries, everything.

dbhattacharyya-codenation replied:
Fixed by [#6d5889](https://github.com/dbhattacharyya-codenation/jcnu18/pull/1/commits/6d5889028a928d4f1ca0ce4dcc15dd2bd151fcf2)

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,403 @@
+                    "MATCH (file:File)-[:DEFINE]->(class) " +
+                    "RETURN id(method) AS methodId, method.modifiers AS modifiers, method.file AS file, method.line AS line, method.col AS col, id(file) AS fileId";
+        }
+        return null;
```
Returning null is usually discouraged unless "null" is somehow a valid response from your function, because NullPointerExceptions are hard to catch until you run into them. In this case, "null" doesn't make sense as a response from getQueryString. Either throw a relevant checked exception or worst case, return "".

dbhattacharyya-codenation replied:
Fixed by [#d0a2f6](https://github.com/dbhattacharyya-codenation/jcnu18/pull/1/commits/d0a2f61edc196dbe2f8a8db3092ef645a279d53c)

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,403 @@
+        return null;
+    }
+
+    // helper method to return query string for modifier order check
```
These comments make your code pretty readable, but adding comments that explain "what"s are considered a code smell. Ideally, you should be extracting every such block that you feel is necessary to label into a new method with a relevant name, so the comment becomes necessary. Add Javadocs to these new methods to explain any "why"s or corner cases if necessary.
Intellij IDEs offer extract method refactoring out of the box.

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,403 @@
+            updatedModifiers += "protected, ";
+        }
+
+        if (currentModifiers.contains("abstract")) {
```
While the cases from here on are very readable, they make the method very long as well. Maybe construct a list with the modifiers in sequence and iterate through it?

dbhattacharyya-codenation replied:
Fixed by [#35cd18](https://github.com/dbhattacharyya-codenation/jcnu18/pull/1/commits/35cd180e25cdb7fc095f0857bf53f6829c5d50d7)

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,403 @@
+            updatedModifiers += "strictfp, ";
+        }
+
+        if (updatedModifiers.length() > 2) {
```
While this condition works as required, it isn't clear why this condition was chosen. Perhaps StringUtils.isNotBlank() from the Apache commons library fits better?

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,403 @@
+            session.close();
+            driver.close();
+        }
+        catch (Exception e) {
```
Do not catch the generic Exception unless you are sure that's what you want, because this catches even RuntimeExceptions like memory-related ones, which you usually don't want to mess with.

Your IDE should help you with what exceptions are actually thrown in that block of code.

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,403 @@
+        // Check for valid fixId
+        if (minifixes.isEmpty()) {
+            return gson.toJson(
+                    new GetResponse(400, "FixId not found")
```
nitpick: 404 is a better code than 400 here.

dbhattacharyya-codenation replied:
Fixed by [#867608](https://github.com/dbhattacharyya-codenation/jcnu18/pull/1/commits/867608310ff083d6560e5db9a4485541d9becb18)

coderick14 committed:
Extract request and response classes to helper package

coderick14 committed:
Extract constants into a package

coderick14 committed:
Add modifiers to a List to reduce method size

coderick14 committed:
Return 404 instead of 400 for fixId not found

coderick14 committed:
Throw InvalidIssueIdException instead of returning null

coderick14 committed:
Move Cypher queries to constants package

pranavram-cn commented on .env:
```diff
@@ -0,0 +1,2 @@
+DB_NAME=miniautofix
+DB_PASS=superstrongpassword
```
Kudos.

pranavram-cn commented on build.gradle:
```diff
@@ -0,0 +1,44 @@
+apply plugin: 'io.spring.dependency-management'
+
+dependencies {
+    compile 'org.springframework.boot:spring-boot-starter-web'
```
Try and stick to a single format for all your dependencies.

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,231 @@
+
+    // helper method to return query string for modifier order check
+    private String getModifierFixQuery(Integer methodId, String updatedModifier) {
+        return String.format("MATCH (m:MethodDeclaration) WHERE id(m) = %d SET m.modifiers = \"%s\"", methodId, updatedModifier);
```
Extract this constant as well

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,231 @@
+            }
+        }
+
+        if (updatedModifiers.length() > 2) {
```
This check still needs to change

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,231 @@
+        }
+
+        updatedModifiers = "[" + updatedModifiers + "]";
+        return updatedModifiers;
```
nitpick: you could just return it one statement earlier

pranavram-cn commented on src/main/java/simplefixer/controller/MinifixController.java:
```diff
@@ -0,0 +1,231 @@
+    private IssueTypeRepository issueTypeRepository;
+
+    @PostMapping(value = "/fix", produces = "application/json")
+    public @ResponseBody String handleIssues(@RequestParam Integer issueId, @RequestParam String sandBoxURL) {
```
This method is extremely long. Ideally, no method should have to be over about 10 statements long- every method must have a clear focus of what it achieves. Break its function down into high-level subtasks and make each of them methods. Continue the same for the subtasks.

pranavram-cn commented on src/main/java/simplefixer/helper/IssueResponse.java:
```diff
@@ -0,0 +1,21 @@
+    private Integer columnNumber;
+    private Boolean isFixed;
+
+    public IssueResponse(Integer id, Integer issueTypeId, String issueTypeDesc,String fileName, Integer lineNumber,Integer columnNumber, Boolean isFixed) {
```
When the number of args to the constructor exceeds 3/4, it's a good idea to use the builder pattern. Use Lombok- it greatly reduces the code you have to write to implement it

pranavram-cn reviewed (changes requested):
When you fix more than one comment, you need not make a new commit for each fix- You could just add a multiline commit message that enumerates all the fixes in that commit

coderick14 committed:
Make requested changes

- Use uniform dependency format in build.gradle
- Use StringUtils.isEmpty() to check for empty string
- Extract getFixModifierQuery to CypherQuery package

pranavram-cn commented on src/main/java/simplefixer/constant/CypherQuery.java:
```diff
@@ -15,4 +15,7 @@
             "MATCH (file:File)-[:DEFINE]->(class) " +
             "RETURN id(method) AS methodId, method.modifiers AS modifiers, method.file AS file, method.line AS line, method.col AS col, id(file) AS fileId";
 
+    public static final String GET_MODIFIER_FIX_QUERY(Integer methodId, String updatedModifier) {
```
Java's naming convention follows camelCase for all methods. For a method "final" simply means it cannot be overridden.
